portal.go

```diff
@@ -2040,29 +2109,167 @@ func (portal *Portal) updateAvatarInRoom(ctx context.Context, sender *Puppet) {
+			if user == nil {
+				continue
+			}
+			mxid = user.MXID
+			membership := currentMembers[mxid]
+			err = nil
+			if membership == event.MembershipJoin || membership == event.MembershipBan {
+				_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipLeave, "")
+			}
+			if membership != event.MembershipInvite {
+				_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipInvite, "")
+			}
 			if err != nil {
-				log.Err(err).Stringer("signal_user_id", member.UserID).Msg("Failed to ensure user is joined to portal")
+				log.Warn().Stringer("mxid", mxid).Msg("Couldn't change membership to invite")
+			}
+			userIDs[mxid] = ((int)(pendingMember.Role) >> 1) * 50
+			delete(currentMembers, mxid)
+		}
+	}
+	for _, requestingMember := range info.RequestingMembers {
+		puppet := portal.bridge.GetPuppetBySignalID(requestingMember.UserID)
+		if puppet == nil {
+			log.Warn().Stringer("signal_user_id", requestingMember.UserID).Msg("Couldn't get puppet for group member")
+			continue
+		}
+		mxid := puppet.IntentFor(portal).UserID
+		membership := currentMembers[mxid]
+		var err error
+		if membership == event.MembershipJoin || membership == event.MembershipBan {
+			_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipLeave, "")
+		}
+		if membership != event.MembershipKnock {
+			_, err = puppet.IntentFor(portal).SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipKnock, "")
+		}
+		if err != nil {
+			log.Warn().Stringer("mxid", mxid).Msg("Couldn't change membership to knock")
+		}
+		delete(currentMembers, mxid)
+	}
+	for _, bannedMember := range info.BannedMembers {
+		puppet := portal.bridge.GetPuppetBySignalID(bannedMember.UserID)
+		if puppet == nil {
+			log.Warn().Stringer("signal_user_id", bannedMember.UserID).Msg("Couldn't get puppet for group member")
+			continue
+		}
+		mxid := puppet.IntentFor(portal).UserID
+		var err error
+		if currentMembers[mxid] != event.MembershipBan {
+			_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipBan, "")
+		}
+		if err != nil {
+			log.Warn().Stringer("mxid", mxid).Msg("Couldn't change membership to ban")
+		}
+		delete(currentMembers, mxid)
+		if puppet.customIntent == nil {
+			user := portal.bridge.GetUserBySignalID(bannedMember.UserID)
+			if user == nil {
+				continue
+			}
+			mxid = user.MXID
+			err = nil
+			if currentMembers[mxid] != event.MembershipBan {
+				_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipBan, "")
+			}
+			if err != nil {
+				log.Warn().Stringer("mxid", mxid).Msg("Couldn't change membership to ban")
+			}
+			delete(currentMembers, mxid)
+		}
+	}
+	for mxid, membership := range currentMembers {
+		if membership == event.MembershipLeave {
+			continue
+		}
+		user := portal.bridge.GetUserByMXIDIfExists(mxid)
+		if user != nil {
```

Comment:
When getting by mxid it's more efficient to try puppet first and then user, because determining if a mxid is a bridge ghost doesn't require a database query